### DISCUSSION TITLE
NORM point-to-point between two processes on the same machine?

### DISCUSSION THREAD

**Discussion by stevek138** [USNavalResearchLaboratory/norm#39] [Q&A] [answered]
If there are two separate processes on two separate machines, and NORM JNI is used to perform a point-to-point connection, the following works:

1. First machine (192.168.91.131):

	NormInstance instance = new NormInstance();
	String addr = "192.168.91.132";
	int port = 9950;
	NormSession session = instance.createSession(addr, port, NormNode.NORM_NODE_ANY);
	session.startSender(...);
	subsequent code to send file or stream

2. Second machine (192.168.1.132):

	NormInstance instance = new NormInstance();
	String addr = "192.168.91.131";
	int port = 9950;
	NormSession session = instance.createSession(addr, port, NormNode.NORM_NODE_ANY);
	session.startReceiver(...);
	subsequent code to receive file or stream

With this, the send works and data is transferred.

If these two separate processes are running on the same machine, say 192.168.91.131, each can't perform a "NormSession session = instance.createSession(addr, port, NormNode.NORM_NODE_ANY);" with the same ip addr and port because the second execution of createSession() results in something like:

Proto Error: ProtoSocket::Bind(9950) bind() error: Address already in use
Proto Fatal: NormSession::Open() error: rx_socket.Bind() error
java.io.IOException: Failed to start sender

The NORM JNI does not include other interface options for initiating a session from the NormInstance class. As such, can NORM JNI be used to create a point-to-point connection between two processes on the same machine/ip, or is this a limitation on the point-to-point implementation via JNI?

**Comment by marygilliland** [comment]
I have only used the C++ API for multicast applications, but in the C++ API there is the NormSetRXPortReuse() call to allow the sender and receiver to run on the same host.  Perhaps there is the same in the Java API?

**Comment by bebopagogo** [comment, marked as answer]
As @marygilliland points out, when running multiple NORM applications on the same machine for the same multicast address/port, you need to use NormSetRxPortReuse() so multiple sockets on the same machine can bind to the same port.  Note this only works for IP multicast addresses due to the way networks stacks are implemented on some machines (the NormSocket API extension that is in progress does use connect() for UDP sockets when a connection-oriented paradigm makes sense and the tighter binding there does allow multiple unicast UDP connections in that case, but I digress)

The JNI API does have the NormSetRxPortReuse() call.  However, while the C++ "norm/examples/normMsgr.cpp" does invoke that option, the corresponding "norm/examples/java/NormMsgr.java" does not.  If you examine the normMsgr.cpp code you can probably see how that should be applied.  I will make an effort to update the NormMsgr.java.   Note that you may also have to enable multicast "loopback" for it to work as well and there is a NORM API for that, too.  That is where it will be good for me to update the NormMsgr.java to make sure it is complete.

The NormMgr.java example is the best Java example in the code base.  There are 3 example applications I have been focused on with the intention of supporting C/C++, Java, and Python equivalents:

normMsgr - uses NORM_OBJECT_DATA as a "message" transport example

normStreamer - uses NORM_OBJECT_STREAM for either "byte-stream" or "message-stream" operation.

normCast - uses NORM_OBJECT_FILE to illustrate file transfer / broadcast

I have not yet created the Java or Python for the latter two, but one can use the C/C++ code as a guide in the meantime along with the NormMsgr.java or normMsgr.py that illustrate using the API in a working multi-threaded application.

**Reply by stevek138** [reply]
Thanks to both @marygilliland and @bebopagogo for the quick and detailed responses. This is very appreciated and helpful. I'm looking into all of this now. If I could just get one quick clarification... does this work for point-to-point or only for multicast capabilities? An I interpreting the "Note this only works for IP multicast addresses..." comment corrently in that there is a NormSocket API extension that will be available in the future, but isn't available at the moment, that will eventually provide this capability for UDP sockets/unicast/point-to-point? In short, it seems that I can't do this on point-to-point yet, correct?

Addendum: The reference to NormSetRxPortReuse() in normMsgr.cpp is only executed for multicast.

**Reply by bebopagogo** [reply]
Port reuse does _not_ work for connectionless unicast UDP sockets.  The way the kernels on various operating systems work is that for received multicast packets when port reuse is enabled, the kernel exhaustively searches for matching socket bindings and gives a copy of the datagram to each matching socket.  But for received unicast packets, it provides only the _first_ matching socket it finds a copy of the datagram.  On Linux (and maybe BSD) this can change and my experience has been the first match seems to correlate to the socket that last sent a packet.  I suspect that is because of some underlying data structure ends up promoting an "active" socket forward in the list (or whatever) ...

Yes - NORM works for unicast.  The NormSocket API extension does manipulate UDP socket bindings and use the system connect() call so that a "connection-oriented" NORM client/server application can work just like a TCP server where the server listens on a well-known port and can NormAccept multiple connections from multiple clients.
